File: src/main/java/com/wxl/dyttcrawler/core/Crawler.java
package com.wxl.dyttcrawler.core;

import com.google.common.util.concurrent.ThreadFactoryBuilder;
import com.wxl.dyttcrawler.scheduler.BatchScheduler;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.collections.CollectionUtils;
import org.apache.commons.lang3.SerializationUtils;
import org.assertj.core.util.Lists;
import org.springframework.util.Assert;
import us.codecraft.webmagic.Page;
import us.codecraft.webmagic.Request;
import us.codecraft.webmagic.Site;
import us.codecraft.webmagic.Task;
import us.codecraft.webmagic.downloader.Downloader;
import us.codecraft.webmagic.pipeline.ConsolePipeline;
import us.codecraft.webmagic.pipeline.Pipeline;
import us.codecraft.webmagic.processor.PageProcessor;
import us.codecraft.webmagic.scheduler.Scheduler;

import java.io.Closeable;
import java.io.IOException;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.UUID;
import java.util.concurrent.*;
import java.util.concurrent.atomic.AtomicLong;
import java.util.concurrent.atomic.AtomicReference;

/**
 * Create by wuxingle on 2020/6/7
 * 爬虫门面
 *
 * @see us.codecraft.webmagic.Spider
 */
@Slf4j
public class Crawler implements Closeable, Task {

    /**
     * 状态
     */
    public enum Status {
        INIT(0), RUNNING(1), STOPPING(2), STOPPED(3);

        Status(int value) {
            this.value = value;
        }

        private int value;

        int getValue() {
            return value;
        }
    }


    /**
     * 任务uuid
     */
    private String taskId;

    /**
     * 页面处理
     */
    private PageProcessor pageProcessor;

    /**
     * 页面下载
     */
    private Downloader downloader;

    /**
     * 结果处理
     */
    private List<Pipeline> pipelines;

    /**
     * 任务队列
     */
    private Scheduler scheduler;

    /**
     * 执行线程池
     */
    private ThreadPool threadPool;

    /**
     * 开始的请求
     */
    private List<Request> startRequests;

    /**
     * 监听器
     */
    private List<CrawlerListener> crawlerListeners = new CopyOnWriteArrayList<>();

    /**
     * 完成时是否结束
     */
    private boolean exitWhenComplete;


    private Site site;

    /****************************************内部状态*****************************************************/

    /**
     * 等待新url的检查间隔
     * 30s
     */
    private static final int WAIT_NEW_URL_CHECK_TIME = 30 * 1000;

    /**
     * 状态
     */
    private AtomicReference<Status> stat = new AtomicReference<>(Status.INIT);


    /**
     * 页面处理计数
     */
    private final AtomicLong pageCount = new AtomicLong(0);

    /**
     * 爬虫主线程池
     */
    private ExecutorService crawlerExecutor = new ThreadPoolExecutor(
            0, 1, 10, TimeUnit.MINUTES, new SynchronousQueue<>(),
            new ThreadFactoryBuilder().setNameFormat("crawler-worker-%s").setDaemon(true).build()
    );

    /**
     * 爬虫主线程future
     */
    private volatile Future<?> crawlerFuture;

    /**
     * create a spider with pageProcessor.
     *
     * @param pageProcessor pageProcessor
     */
    Crawler(String taskId,
            PageProcessor pageProcessor,
            Downloader downloader,
            List<Pipeline> pipelines,
            Scheduler scheduler,
            ThreadPool threadPool,
            List<Request> startRequests,
            boolean exitWhenComplete) {
        Assert.notNull(pageProcessor, "page processor can not null");
        Assert.notNull(downloader, "http downloader can not null");
        Assert.notNull(scheduler, "scheduler can not null");
        Assert.isTrue(threadPool != null && !threadPool.isShutDown(), "thread pool must is active");
        Assert.isTrue(CollectionUtils.isNotEmpty(startRequests), "start urls can not empty");
        this.taskId = taskId;
        this.pageProcessor = pageProcessor;
        this.downloader = downloader;
        if (CollectionUtils.isEmpty(pipelines)) {
            this.pipelines = Lists.newArrayList(new ConsolePipeline());
        } else {
            this.pipelines = Lists.newArrayList(pipelines);
        }
        this.scheduler = scheduler;
        this.threadPool = threadPool;
        this.startRequests = Collections.unmodifiableList(startRequests);
        this.exitWhenComplete = exitWhenComplete;
        this.site = pageProcessor.getSite();

    }

    /**
     * 手动调用url
     */
    public void crawl(String url) {
        processRequest(new Request(url));
    }

    /**
     * 异步启动
     */
    public boolean start(long maxCount) {
        // check running
        while (true) {
            Status statNow = stat.get();
            if (statNow == Status.RUNNING) {
                log.info("crawler {} already started!", getUUID());
                return false;
            }
            if (statNow == Status.STOPPING) {
                log.info("crawler {} is stopping!", getUUID());
                return false;
            }
            if (stat.compareAndSet(statNow, Status.RUNNING)) {
                // add first request
                if (CollectionUtils.isNotEmpty(startRequests)) {
                    addRequest(startRequests);
                }
                break;
            }
        }
        crawlerFuture = crawlerExecutor.submit(() -> doCrawlerSpin(maxCount));
        return true;
    }

    public boolean start() {
        return start(Long.MAX_VALUE);
    }

    /**
     * 异步停止
     */
    public boolean stop() {
        return stat.compareAndSet(Status.RUNNING, Status.STOPPING);
    }

    /**
     * 等待执行结束
     */
    public void awaitStopped() throws InterruptedException {
        if (crawlerFuture != null) {
            while (!crawlerFuture.isDone() && !crawlerFuture.isCancelled()) {
                Thread.sleep(5000);
            }
        }
    }

    @Override
    public void close() {
        stop();
        try {
            awaitStopped();
        } catch (InterruptedException e) {
            //ignore
        }
        try {
            threadPool.shutdown();
        } catch (InterruptedException e) {
            //ignore
        }
        for (Pipeline pipeline : pipelines) {
            destroyEach(pipeline);
        }
        destroyEach(pageProcessor);
        destroyEach(scheduler);
        destroyEach(downloader);
    }

    private void destroyEach(Object object) {
        if (object instanceof ExecutorService) {
            ExecutorService executorService = (ExecutorService) object;
            try {
                executorService.shutdown();
                while (!executorService.isTerminated()) {
                    executorService.awaitTermination(5, TimeUnit.SECONDS);
                }
            } catch (InterruptedException e) {
                log.error("threadPoll {} has interrupt", executorService);
            }
        } else if (object instanceof Closeable) {
            try {
                ((Closeable) object).close();
            } catch (IOException e) {
                log.error("close {} has error", object, e);
            }
        }
    }

    /**
     * 爬虫执行逻辑
     *
     * @param maxCount 最大爬取页面数
     */
    private void doCrawlerSpin(long maxCount) {
        log.info("crawler {} started!", getUUID());
        long count = 0;
        long startIndex = pageCount.get();
        while (!Thread.currentThread().isInterrupted()
                && stat.get() == Status.RUNNING
                && count++ < maxCount) {
            final Request request = scheduler.poll(this);
            if (request == null) {
                if (threadPool.getActiveThreadNum() == 0 && exitWhenComplete) {
                    break;
                }
                // wait until new url added
                await(WAIT_NEW_URL_CHECK_TIME);
                count--;
            } else {
                try {
                    threadPool.execute(() -> processRequest(request));
                } catch (InterruptedException e) {
                    log.info("submit process request task is interrupted");
                    addRequest(request);
                    Thread.currentThread().interrupt();
                    break;
                }
            }
        }
        while (!Thread.currentThread().isInterrupted()
                && threadPool.getActiveThreadNum() != 0
                && (pageCount.get() - startIndex < maxCount)) {
            await(5000);
        }

        log.info("crawler {} closed! {} pages downloaded.", getUUID(), pageCount.get());
        stat.set(Status.STOPPED);
    }

    /**
     * wait
     */
    private void await(long timeout) {
        synchronized (this) {
            try {
                wait(timeout);
            } catch (InterruptedException e) {
                log.info("wait is interrupted");
                Thread.currentThread().interrupt();
            }
        }
    }

    /**
     * notify
     */
    private void signalAll() {
        synchronized (this) {
            notifyAll();
        }
    }

    /**
     * 请求处理
     */
    private void processRequest(Request request) {
        try {
            Page page = downloader.download(request, this);
            if (page.isDownloadSuccess()) {
                onDownloadSuccess(request, page);
            } else {
                onDownloaderFail(request);
            }
            notifySuccess(request);
        } catch (Exception e) {
            notifyError(request);
            log.error("process request " + request + " error", e);
        } finally {
            pageCount.incrementAndGet();
            signalAll();
        }
    }

    /**
     * 成功处理
     */
    private void onDownloadSuccess(Request request, Page page) {
        if (site.getAcceptStatCode().contains(page.getStatusCode())) {
            pageProcessor.process(page);
            if (CollectionUtils.isNotEmpty(page.getTargetRequests())) {
                addRequest(page.getTargetRequests());
            }
            if (!page.getResultItems().isSkip()) {
                for (Pipeline pipeline : pipelines) {
                    pipeline.process(page.getResultItems(), this);
                }
            }
        } else {
            log.info("page status code error, page {} , code: {}", request.getUrl(), page.getStatusCode());
        }
        sleep(site.getSleepTime());
    }

    /**
     * 失败处理
     */
    private void onDownloaderFail(Request request) {
        if (site.getCycleRetryTimes() == 0) {
            sleep(site.getSleepTime());
        } else {
            // for cycle retry
            doCycleRetry(request);
        }
    }

    private void doCycleRetry(Request request) {
        Object cycleTriedTimesObject = request.getExtra(Request.CYCLE_TRIED_TIMES);
        if (cycleTriedTimesObject == null) {
            addRequest(SerializationUtils.clone(request).setPriority(0).putExtra(Request.CYCLE_TRIED_TIMES, 1));
        } else {
            int cycleTriedTimes = (Integer) cycleTriedTimesObject;
            cycleTriedTimes++;
            if (cycleTriedTimes < site.getCycleRetryTimes()) {
                addRequest(SerializationUtils.clone(request).setPriority(0).putExtra(Request.CYCLE_TRIED_TIMES, cycleTriedTimes));
            }
        }
        sleep(site.getRetrySleepTime());
    }

    private void sleep(int time) {
        try {
            Thread.sleep(time);
        } catch (InterruptedException e) {
            log.info("crawler sleep is interrupted");
        }
    }

    /**
     * 请求放队列
     */
    private void addRequest(Request request) {
        scheduler.push(request, this);
    }

    private void addRequest(List<Request> requests) {
        if (scheduler instanceof BatchScheduler) {
            ((BatchScheduler) scheduler).push(requests, this);
        } else {
            for (Request request : requests) {
                addRequest(request);
            }
        }
    }

    private void notifyError(Request request) {
        if (!CollectionUtils.isEmpty(crawlerListeners)) {
            for (CrawlerListener crawlerListener : crawlerListeners) {
                try {
                    crawlerListener.onError(request, this);
                } catch (Exception e) {
                    log.error("crawler on error process error:{}", request, e);
                }
            }
        }
    }

    private void notifySuccess(Request request) {
        if (!CollectionUtils.isEmpty(crawlerListeners)) {
            for (CrawlerListener crawlerListener : crawlerListeners) {
                try {
                    crawlerListener.onSuccess(request, this);
                } catch (Exception e) {
                    log.error("crawler on success process error:{}", request, e);
                }
            }
        }
    }

    @Override
    public String getUUID() {
        if (taskId != null) {
            return taskId;
        }
        if (site != null) {
            return site.getDomain();
        }
        taskId = UUID.randomUUID().toString();
        return taskId;
    }


    @Override
    public Site getSite() {
        return site;
    }

    public void addCrawlerListener(CrawlerListener... listeners) {
        this.crawlerListeners.addAll(Arrays.asList(listeners));
    }

    public long getPageCount() {
        return pageCount.get();
    }

    public Status getStatus() {
        return stat.get();
    }
}

File: src/main/kotlin/com/wxl/dyttcrawler/core/Crawler.kt
package com.wxl.dyttcrawler.core

import com.wxl.dyttcrawler.downloader.HttpDownloader
import com.wxl.dyttcrawler.scheduler.BatchScheduler
import kotlinx.coroutines.*
import kotlinx.coroutines.channels.Channel
import org.slf4j.LoggerFactory
import us.codecraft.webmagic.Page
import us.codecraft.webmagic.Request
import us.codecraft.webmagic.Site
import us.codecraft.webmagic.Task
import us.codecraft.webmagic.pipeline.Pipeline
import us.codecraft.webmagic.processor.PageProcessor
import us.codecraft.webmagic.scheduler.Scheduler
import java.io.Closeable
import java.io.IOException
import java.util.*
import java.util.concurrent.*
import java.util.concurrent.atomic.AtomicLong
import java.util.concurrent.atomic.AtomicReference

/**
 * Create by wuxingle on 2021/10/09
 * 爬虫相关类
 */

/**
 * 等待新url的检查间隔
 * 30s
 */
private const val WAIT_NEW_URL_CHECK_TIME = 30 * 1000L

/**
 * 状态类
 */
enum class CrawlerStatus(val value: Int) {
    INIT(0),
    RUNNING(1),
    STOPPING(2),
    STOPPED(3);
}

/**
 * 爬虫启动类
 */
class Crawler private constructor(
    private val taskId: String,
    private val pageProcessor: PageProcessor,
    private val downloader: HttpDownloader,
    private val pipelines: List<Pipeline>,
    private val scheduler: Scheduler,
    private val coroutineScope: CoroutineScope,
    private val startRequests: List<Request>,
    private val exitWhenComplete: Boolean = false,
    private val concurrentNums: Int = 1
) : Closeable, Task {

    /**
     * builder
     */
    class CrawlerBuilder {
        /**
         * 任务uuid
         */
        var taskId: String = UUID.randomUUID().toString()

        /**
         * 页面处理
         */
        lateinit var pageProcessor: PageProcessor

        /**
         * 页面下载
         */
        lateinit var downloader: HttpDownloader

        /**
         * 结果处理
         */
        lateinit var pipelines: List<Pipeline>

        /**
         * 任务队列
         */
        lateinit var scheduler: Scheduler

        /**
         * 协程scope
         */
        lateinit var coroutineScope: CoroutineScope

        /**
         * 开始的请求
         */
        lateinit var startRequests: List<Request>

        /**
         * 完成时是否结束
         */
        var exitWhenComplete: Boolean = false

        /**
         * 并发度
         */
        var concurrentNums: Int = 1

        /**
         * 监听器
         */
        var crawlerListeners: MutableList<CrawlerListener> = mutableListOf()

        fun build(): Crawler {
            val listeners = crawlerListeners.toTypedArray()
            return Crawler(
                taskId, pageProcessor, downloader,
                pipelines, scheduler, coroutineScope,
                startRequests, exitWhenComplete, concurrentNums
            ).apply {
                addCrawlerListener(*listeners)
            }
        }
    }

    /**
     * 监听器
     */
    private val crawlerListeners: MutableList<CrawlerListener> = CopyOnWriteArrayList()

    /**
     * site
     */
    private val site: Site = pageProcessor.site

    /****************************************内部状态*****************************************************/

    /**
     * 状态
     */
    private val stat: AtomicReference<CrawlerStatus> = AtomicReference(CrawlerStatus.INIT)

    /**
     * 页面处理计数
     */
    private val pageCount = AtomicLong(0)

    /**
     * 并发控制channel
     */
    private val concurrentChannel = Channel<Int>(concurrentNums)

    @Volatile
    private var crawlerJob: Job? = null

    companion object {
        private val log = LoggerFactory.getLogger(Crawler::class.java)

        /**
         * 带接收者的函数类型
         */
        inline fun build(block: CrawlerBuilder.() -> Unit): Crawler =
            CrawlerBuilder().apply(block).build()
    }

    /**
     * 手动调用url
     */
    fun crawl(url: String) = processRequest(Request(url))

    /**
     * 异步启动
     */
    fun start(): Boolean {
        while (true) {
            val statNow = stat.get()
            if (statNow == CrawlerStatus.RUNNING) {
                log.info("crawler {} already started!", uuid)
                return false
            }
            if (statNow == CrawlerStatus.STOPPING) {
                log.info("crawler {} is stopping!", uuid)
                return false
            }
            if (stat.compareAndSet(statNow, CrawlerStatus.RUNNING)) {
                if (startRequests.isNotEmpty()) {
                    addRequest(startRequests)
                }
                break
            }
        }

        crawlerJob = newCrawlerJob()
        crawlerJob!!.start()

        return true
    }

    /**
     * 创建新的爬虫任务
     */
    private fun newCrawlerJob(): Job =
        coroutineScope.launch(CoroutineName("cp"), start = CoroutineStart.LAZY) {
            try {
                log.info("crawler {} started!", uuid)
                while (isActive && stat.get() == CrawlerStatus.RUNNING) {
                    val request = scheduler.poll(this@Crawler)
                    if (request == null) {
                        val emptyChildrenJob = coroutineContext.job.children.iterator().hasNext()
                        if (emptyChildrenJob && concurrentChannel.isEmpty && exitWhenComplete) {
                            log.info("crawler {} execute complete by empty request", uuid)
                            break
                        }
                        delay(1000)
                    } else {
                        log.debug("crawler {} wait download: {}", uuid, request.url)
                        concurrentChannel.send(0)

                        launch(CoroutineName("ch")) {
                            try {
                                processRequest(request)
                                delay(site.sleepTime.toLong())
                            } finally {
                                concurrentChannel.receive()
                            }
                        }
                    }
                }

                // 等待处理完成
                for (child in coroutineContext.job.children) {
                    child.join()
                }
                log.info("crawler {} closed! {} pages downloaded.", uuid, pageCount.get())

            } finally {
                stat.set(CrawlerStatus.STOPPED)
            }
        }

    /**
     * 请求处理
     */
    private fun processRequest(request: Request) {
        try {
            val page = downloader.download(request, this@Crawler)
            if (page.isDownloadSuccess) {
                onDownloadSuccess(request, page)
            } else {
                onDownloaderFail(request)
            }
        } catch (e: Exception) {
            notifyError(request, e)
            log.error("process request {} error", request, e)
        } finally {
            pageCount.incrementAndGet()
        }
    }

    /**
     * 异步停止
     */
    fun stop() = stat.compareAndSet(CrawlerStatus.RUNNING, CrawlerStatus.STOPPING)

    /**
     * 等待停止
     */
    fun awaitStopped() = runBlocking {
        crawlerJob?.join()
    }

    /**
     * 添加监听
     */
    fun addCrawlerListener(vararg listeners: CrawlerListener) =
        crawlerListeners.addAll(listeners)

    /**
     * 获取页面处理数
     */
    fun getPageCount(): Long = pageCount.get()

    /**
     * 获取当前状态
     */
    fun getStatus(): CrawlerStatus = stat.get()

    override fun close() {
        stop()
        awaitStopped()
        runBlocking {
            concurrentChannel.close()
        }

        for (pipeline in pipelines) {
            destroyEach(pipeline)
        }
        destroyEach(pageProcessor)
        destroyEach(scheduler)
        destroyEach(downloader)
    }

    private fun destroyEach(obj: Any) {
        if (obj is Closeable) {
            try {
                obj.close()
            } catch (e: IOException) {
                log.error("close {} has error", obj, e)
            }
        }
    }

    /**
     * 成功处理
     */
    private fun onDownloadSuccess(request: Request, page: Page) {
        if (site.acceptStatCode.contains(page.statusCode)) {
            pageProcessor.process(page)
            if (page.targetRequests.isNotEmpty()) {
                addRequest(page.targetRequests)
            }
            if (!page.resultItems.isSkip) {
                for (pipeline in pipelines) {
                    pipeline.process(page.resultItems, this)
                }
            }
            notifySuccess(request)
        } else {
            log.info("page status code error, page {} , code: {}", request.url, page.statusCode)
            notifyError(request)
        }
    }

    /**
     * 失败处理
     */
    private fun onDownloaderFail(request: Request) {
        log.info("page download fail, page {}", request.url)
        notifyError(request)
    }

    /**
     * 请求放队列
     */
    private fun addRequest(request: Request) {
        scheduler.push(request, this)
    }

    private fun addRequest(requests: List<Request>) {
        if (scheduler is BatchScheduler) {
            scheduler.push(requests, this)
        } else {
            requests.forEach { addRequest(it) }
        }
    }

    private fun notifySuccess(request: Request) {
        for (crawlerListener in crawlerListeners) {
            try {
                crawlerListener.onSuccess(request, this)
            } catch (e: Exception) {
                log.error("crawler on success process error:{}", request, e)
            }
        }
    }

    private fun notifyError(request: Request, e: Exception? = null) {
        for (crawlerListener in crawlerListeners) {
            try {
                crawlerListener.onError(request, this, e)
            } catch (e: Exception) {
                log.error("crawler on error process error:{}", request, e)
            }
        }
    }


    override fun getUUID(): String = taskId

    override fun getSite(): Site = site


}


